File: common/src/main/java/com/imsproject/utils/dataAccess/abstracts/DAO.java
package com.imsproject.utils.dataAccess.abstracts;

import com.imsproject.utils.exceptions.DaoException;

import java.util.List;

/**
 * Data Access Object interface
 * @param <T> the type of the object
 * @param <PK> the type of the primary key
 */
public interface DAO<T, PK extends PrimaryKey> {

    T select(PK key) throws DaoException;

    List<T> selectAll() throws DaoException;

    void insert(T object) throws DaoException;

    void insertAll(List<T> objects) throws DaoException;

    void update(T object) throws DaoException;

    void updateAll(List<T> objects) throws DaoException;

    void delete(PK object) throws DaoException;

    void deleteAll(List<PK> objects) throws DaoException;

    boolean exists(PK object) throws DaoException;
}

File: common/src/main/java/com/imsproject/common/dataAccess/abstracts/DAO.kt
package com.imsproject.common.dataAccess.abstracts

import com.imsproject.common.dataAccess.DaoException

/**
 * Data Access Object interface
 */
interface DAO<T, PK : PrimaryKey> {
    @Throws(DaoException::class)
    fun select(key: PK, transactionId : String? = null): T

    @Throws(DaoException::class)
    fun selectAll(transactionId : String? = null): List<T>

    @Throws(DaoException::class)
    fun selectAll(keys: List<PK>, transactionId : String? = null): List<T>

    @Throws(DaoException::class)
    fun insert(obj: T, transactionId : String? = null) : Int

    @Throws(DaoException::class)
    fun insertAll(objs: List<T>, transactionId : String? = null): List<Int>

    @Throws(DaoException::class)
    fun update(obj: T, transactionId : String? = null)

    @Throws(DaoException::class)
    fun updateAll(objs: List<T>, transactionId : String? = null)

    @Throws(DaoException::class)
    fun delete(key: PK, transactionId : String? = null)

    @Throws(DaoException::class)
    fun deleteAll(keys: List<PK>, transactionId : String? = null)

    @Throws(DaoException::class)
    fun exists(key: PK, transactionId : String? = null): Boolean
}
